src/appservice/Appservice.ts

```diff
@@ -958,16 +958,26 @@ export class Appservice extends EventEmitter {
         }
 
         let responded = false;
-        this.emit("query.key_claim", req.body, async (result: MSC3983KeyClaimResponse | Promise<MSC3983KeyClaimResponse> | undefined | null) => {
-            if (result?.then) result = await result;
-            if (!result) {
-                res.status(404).json({ errcode: "M_UNRECOGNIZED", error: "Endpoint not implemented" });
-                responded = true;
-                return;
-            }
-
-            res.status(200).json(result);
+        this.emit("query.key_claim", req.body, (result: MSC3983KeyClaimResponse | Promise<MSC3983KeyClaimResponse> | undefined | null) => {
             responded = true;
+
+            const handleResult = (result2: MSC3983KeyClaimResponse) => {
+                if (!result2) {
+                    res.status(404).json({ errcode: "M_UNRECOGNIZED", error: "Endpoint not implemented" });
+                    return;
+                }
+
+                res.status(200).json(result2);
+            };
+
+            if ((result as Promise<MSC3983KeyClaimResponse>)?.then) {
```

Comment:
If I was being picky, I'd just wrap `result` in a `Promise.resolve(result)` to save the if condition...but since it's not a functional change I won't block.